### ISSUE TITLE
Implement DurationFormat

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#801]
ECMA-402 will soon include [DurationFormat](https://github.com/tc39/proposal-intl-duration-format), a formatter that supports durations.

Under the hood, DurationFormat uses a ListFormat (#800) to compose measurement units (#275) in most cases, with a special code path for the "digital" format based on [UTS 35](https://unicode.org/reports/tr35/tr35-general.html#durationUnit).

This is therefore blocked on #275 and #800.

Things to keep in mind:

- If implemented correctly, DurationFormat should require NumberFormat but not DateTimeFormat.
- Look carefully at the ECMA-402 spec for how to handle fallbacks and edge cases.  Test thoroughly!

**Comment by DesmondWillowbrook** [comment]
I'm interested in working on this for GSoC, but I'm not sure of the right place to discuss this proposal. Should it be in this issue, on Github Discussions, or somewhere else? (I've looked at CONTRIBUTING.md, but it doesn't seem to have any advice on this)

**Comment by Manishearth** [comment]
Yes, you can ask questions here or on GitHub Discussions. We also have gsoc@unicode.org

**Comment by sffc** [comment]
Questions specifically about DurationFormat are on-topic for this thread.

**Comment by DesmondWillowbrook** [comment]
Are the attached issues (#275 and #800) considered as part of the scope for the GSoC proposal? (They don't seem to have been completely resolved yet)

**Comment by DesmondWillowbrook** [comment]
Would [gsoc@unicode.org](mailto:gsoc@unicode.org) be the right place to discuss GSoC proposals for this idea?

**Comment by DesmondWillowbrook** [comment]
ListFormat functionality seems to exist in ICU4X, but I wasn't able to find any mention of unit formatting in the codebase (only discussion in the linked issue)

I'm starting to draft my proposal; would you say that implementing unit formatting is part of the scope?

**Comment by DesmondWillowbrook** [comment]
I've found [this CLDR data](https://www.unicode.org/cldr/charts/44/by_type/units.duration.html) for duration units. Here's my understanding after reviewing [the DurationFormat proposal](https://github.com/tc39/proposal-intl-duration-format):
- We format duration units together with ListFormat.
- The numbers need to be displayed with units, which internally use `FixedDecimalFormatter` for number formatting. Displaying units using `FixedDecimalFormatter` is not built yet. For duration formatting, we may only need #2193.

**Comment by sffc** [comment]
@DesmondWillowbrook Thanks for looking into this!

Yes, you found the CLDR data for the long, short, and narrow styles. In addition, for the numeric style, you'll need this data:

https://github.com/unicode-org/cldr/blob/c7e39f13da2bfbaf58c1447d610627511c7a6549/common/main/root.xml#L5650C4-L5650C16

ListFormatter is implemented already, and the parts of FixedDecimalFormatter that you need for duration formatting are also implemented. You can also use the new `icu_pattern` crate in order to help with your data handling.

We may eventually want to implement data sharing between the duration formatter and the single units formatter, but to get started, these can be implemented as separate features. You do not need to wait for single unit formatter to be implemented.

You should start your work in the `icu_experimental` crate. Moving it to `icu_duration` is the last step after it is ready to ship.
